### PR TITLE
Ignore packet loss when the keys are not available

### LINKED ISSUE #1967: Loss events for reordered 0- and 0.5-RTT data
In the event that 0-or 0.5-RTT data arrives before the cryptographic handshake packets that would normally precede it.  That is, the Initial containing ClientHello might arrive after a few 0-RTT packets, or the 0.5-RTT packets might arrive before the server flight is completely received.

In that case, those packets will be declared lost.  That's fine from a loss-recovery perspective.  However, the congestion controller will treat this as a loss event.  We want to ensure that loss events are driven only by congestion.

Perhaps the congestion controller shouldn't trigger a loss event on packets that might be dropped in this way.  If we do this, we need to be careful about definition.

### PR DESCRIPTION
**Pull request by ianswett**
Ignore the loss of Handshake, 0-RTT, and 1-RTT packets if the peer doesn't have the keys to process them.

Also unifies the text on discarding keys and no longer implies that 0-RTT has it's own packet number space.

Fixes #1967  Replaces #2028

### CODE REVIEW

File: draft-ietf-quic-recovery.md [hunk 1]
```diff
@@ -427,15 +427,33 @@ packet is received.  The client MAY use this value to seed the RTT estimator for
 a subsequent connection attempt to the server.
 
 
-#### Discarding Initial State {#discard-initial}
+#### Discarding Keys and Packet State {#discarding-packets}
+
+When keys are discarded, recovery state for all in-flight packets sent with
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md]: I think that we want to use "packet protection keys" on the first mention here.

```suggestion
When packet protection keys are discarded (see Section 4.9 of {{QUIC-TLS}}), recovery state for all in-flight packets sent with
```

File: draft-ietf-quic-recovery.md [hunk 2]
```diff
@@ -427,15 +427,33 @@ packet is received.  The client MAY use this value to seed the RTT estimator for
...
+When 0-RTT is rejected, recovery state for all in-flight 0-RTT packets is
+discarded.
+
+If a server accepts 0-RTT, but does not buffer 0-RTT packets that arrive
+before Initial packets, early 0-RTT packets will be declared lost, but that
+is expected to be infrequent.
+
+#### Ignoring Loss of Undecryptable Packets
 
+During the handshake, the keys to remove packet protection may not yet be
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md]:
```suggestion
During the handshake, some packet protection keys might not be
```

File: draft-ietf-quic-recovery.md [hunk 3]
```diff
@@ -427,15 +427,33 @@ packet is received.  The client MAY use this value to seed the RTT estimator for
...
+available when the packet arrives. In particular, Handshake and 0-RTT packets
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md]:
```suggestion
available when a packet arrives. In particular, Handshake and 0-RTT packets
```

File: draft-ietf-quic-recovery.md [hunk 4]
```diff
@@ -427,15 +427,33 @@ packet is received.  The client MAY use this value to seed the RTT estimator for
...
+will not be processable until the Initial flight arrives and 1-RTT packets
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md]:
```suggestion
can't be processed until the Initial flight arrives and 1-RTT packets
```

File: draft-ietf-quic-recovery.md [hunk 5]
```diff
@@ -427,15 +427,33 @@ packet is received.  The client MAY use this value to seed the RTT estimator for
...
+will not be processable until the handshake completes.  Implementations MAY
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md]:
```suggestion
can't be processed until the handshake completes.  Endpoints MAY
```

File: draft-ietf-quic-recovery.md [hunk 6]
```diff
@@ -427,15 +445,34 @@ packet is received.  The client MAY use this value to seed the RTT estimator for
...
+point, recovery state for all in-flight Initial packets is discarded.
 
+When 0-RTT is rejected, recovery state for all in-flight 0-RTT packets is
+discarded.
+
+If a server accepts 0-RTT, but does not buffer 0-RTT packets that arrive
+before Initial packets, early 0-RTT packets will be declared lost, but that
+is expected to be infrequent.
+
+#### Ignoring Loss of Undecryptable Packets
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: This doesn't belong here. You want this in the CC section.

File: draft-ietf-quic-recovery.md [hunk 7]
```diff
@@ -427,15 +445,34 @@ packet is received.  The client MAY use this value to seed the RTT estimator for
 a subsequent connection attempt to the server.
 
 
-#### Discarding Initial State {#discard-initial}
+#### Discarding Keys and Packet State {#discarding-packets}
+
+When packet protection keys are discarded (see Section 4.9 of {{QUIC-TLS}}),
+recovery state for all in-flight packets sent with those keys is discarded.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: I think a slight restructure might help make this text clearer. I've also rephrased some.

"When packet protection keys are discarded (see Section 4.9 of {{QUIC-TLS}}),
all packets that were sent with those keys can no longer be acknowledged because
their acknowledgements cannot be processed anymore. The sender considers them
no longer in flight.  That is, the sender SHOULD discard all recovery state associated
with those packets and MUST remove them from the count of bytes in flight.

As described in Section 17.5.1 of {{QUIC-TRANSPORT}}, endpoints stop sending and
receiving Initial packets once they start exchanging Handshake packets.  At this
point, recovery state for all in-flight Initial packets is discarded.

When 0-RTT is rejected, recovery state for all in-flight 0-RTT packets is
discarded.

If a server accepts 0-RTT, but does not buffer 0-RTT packets that arrive
before Initial packets, early 0-RTT packets will be declared lost, but that
is expected to be infrequent.

It is expected that keys are discarded after packets encrypted with them would be
acknowledged or declared lost, with the exception of Initial secrets, which might
destroyed earlier."

File: draft-ietf-quic-recovery.md [hunk 8]
```diff
@@ -427,15 +445,34 @@ packet is received.  The client MAY use this value to seed the RTT estimator for
...
+The packets are removed from the count of bytes in flight and no
+acknowledgements or loss events will occur for those packets.  Note that it
+is expected that keys are discarded after those packets would be declared lost,
+but Initial secrets are destroyed earlier.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: I don't understand this last sentence. Can you rephrase?

File: draft-ietf-quic-recovery.md [hunk 9]
```diff
@@ -427,15 +445,34 @@ packet is received.  The client MAY use this value to seed the RTT estimator for
...
+
+If a server accepts 0-RTT, but does not buffer 0-RTT packets that arrive
+before Initial packets, early 0-RTT packets will be declared lost, but that
+is expected to be infrequent.
+
+#### Ignoring Loss of Undecryptable Packets
+
+During the handshake, some packet protection keys might not be
+available when a packet arrives. In particular, Handshake and 0-RTT packets
+can't be processed until the Initial flight arrives and 1-RTT packets
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
cannot be processed until the Initial packets arrive, and 1-RTT packets
```

File: draft-ietf-quic-recovery.md [hunk 10]
```diff
@@ -427,15 +445,34 @@ packet is received.  The client MAY use this value to seed the RTT estimator for
...
+can't be processed until the handshake completes.  Endpoints MAY
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
cannot be processed until the handshake completes.  Endpoints MAY
```

File: draft-ietf-quic-recovery.md [hunk 11]
```diff
@@ -427,15 +445,34 @@ packet is received.  The client MAY use this value to seed the RTT estimator for
...
+ignore the loss of Handshake, 0-RTT, and 1-RTT packets that arrive before the
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
ignore the loss of Handshake, 0-RTT, and 1-RTT packets that might arrive before the
```

File: draft-ietf-quic-recovery.md (hunk 8 above)
Comment (kazuho) [reply draft-ietf-quic-recovery.md]: IIUC this refers to the provision in https://quicwg.org/base-drafts/draft-ietf-quic-tls.html#rfc.section.4.10. Maybe it's worth adding a reference.

File: draft-ietf-quic-recovery.md [hunk 12]
```diff
@@ -945,6 +973,14 @@ The recovery period limits congestion window reduction to once per round trip.
 During recovery, the congestion window remains unchanged irrespective of new
 losses or increases in the ECN-CE counter.
 
+#### Ignoring Loss of Undecryptable Packets
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md]: I think that you want H2 here, definitely not H4.

```suggestion
## Ignoring Loss of Undecryptable Packets
```

File: draft-ietf-quic-recovery.md [hunk 13]
```diff
@@ -44,6 +44,24 @@ normative:
...
+      -
+        ins: M. Thomson
+        name: Martin Thomson
+        org: Mozilla
+        role: editor
+      -
+        ins: S. Turner
+        name: Sean Turner
+        org: sn3rd
+        email: sean@sn3rd.com
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md]:
```suggestion
```

**Comment by janaiyengar** [comment]
merging this one as well to get this in -18.
